components/dependency_injection/autowiring.rst

```diff
@@ -0,0 +1,411 @@
+                <service id="twitter_client" class="AppBundle\TwitterClient" autowire="true" />
+            </services>
+        </services>
+
+    .. code-block:: php
+
+        use Symfony\Component\DependencyInjection\Definition;
+
+        // ...
+        $definition = new Definition('AppBundle\TwitterClient');
+        $definition->setAutowired(true);
+
+        $container->setDefinition('twitter_client', $definition);
+
+The autowiring subsystem will parse the constructor of the ``TwitterClient`` class
+and detects its dependencies that way. Here it will find and fill the need for
+an instance of a ``Rot13Transformer``.
+
+If an existing service definition (and only one – see below) is of the needed type,
+it will inject it. Here it’s not the case, but the subsystem is smart enough to
+automatically register a private service for the Rot13Transformer class and set
+it as first argument of the ``twitter_client`` service. Again, it can work only
+if there is one class of the given type. If there are several classes of the same
+type, you must fallback to the explicit service definition or register a default
+implementation.
+
+As you can see, the autowiring feature drastically reduces the amount of configuration
+required to define a service. No more arguments section! It also makes it easy
+to change the dependencies of the ``TwitterClient`` class: just add or remove typehinted
+arguments in the constructor and you’re done. There is no need anymore to search
+and edit related service definitions.
+
+Here is a typical controller using the ``twitter_client`` service::
+
+    // src/AppBundle/Controller/DefaultController.php
+
+    namespace AppBundle\Controller;
+
+    use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+    use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
+    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+    class DefaultController extends Controller
+    {
+        /**
+         * @Route("/tweet")
+         * @Method("POST")
+         */
+        public function tweetAction(Request $request)
+        {
+            $user = $request->request->get('user');
+            $key = $request->request->get('key');
+            $status = $request->request->get('status');
+
+            if (!$user || !$key || !$status) {
+                throw new BadRequestHttpException();
+            }
+
+            $this->get('twitter_client')->tweetInRot13($user, $key, $status);
+
+            return new Response('OK');
+        }
+    }
+
+You can give a try to the API with ``curl``::
+
+    curl -d "user=kevin&key=ABCD&status=Salut" http://localhost:8000/tweet
+
+It should return ``OK``.
+
+Working with Interfaces
+-----------------------
+
+This is nice but when the application grows, it’s recommended to code against abstractions
```

Comment:
(L154 - 156) You might also find yourself using abstractions instead of implementations (especially in grown applications) as it allows to easily replace some dependencies without modifying the class depending of them.